hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/SourceBuilder.java

```diff
@@ -0,0 +1,514 @@
+         * method takes two argument: the item and the timestamp in milliseconds.
+         * <p>
+         * On any given invocation the function may also choose not to add
+         * any items. Jet will automatically employ an exponential backoff strategy
+         * to avoid calling your function in a tight loop during a period when
+         * there's no data ready to be emitted.
+         *
+         * @param fillBufferFn function that fills the buffer with source data
+         * @param <T_NEW> type of the emitted items
+         * @return this builder with the item type reset to the one inferred from
+         *         {@code fillBufferFn}
+         */
+        @SuppressWarnings("unchecked")
+        public <T_NEW> TimestampedStream<T_NEW> fillBufferFn(
+                @Nonnull DistributedBiConsumer<? super S, ? super TimestampedSourceBuffer<T_NEW>> fillBufferFn
+        ) {
+            TimestampedStream<T_NEW> newThis = (TimestampedStream<T_NEW>) this;
+            newThis.fillBufferFn = fillBufferFn;
+            return newThis;
+        }
+
+        /**
+         * Sets the function that Jet will call when cleaning up after a job has
+         * ended. It gives you the opportunity to release any resources held by the
+         * state object.
+         */
+        public TimestampedStream<T> destroyFn(@Nonnull DistributedConsumer<? super S> pDestroyFn) {
+            mDestroyFn = pDestroyFn;
+            return this;
+        }
+
+        /**
+         * Declares that you're creating a distributed source. On each member of
+         * the cluster Jet will create as many workers as you specify with the
+         * {@code preferredLocalParallelism} parameter. If you call this, you must
+         * ensure that all the source workers are coordinated and not emitting
+         * duplicated data. The {@code createFn} can consult {@link Processor.Context#totalParallelism()
+         * procContext.totalParallelism()} and {@link Processor.Context#globalProcessorIndex()
+         * procContext.globalProcessorIndex()}. Jet calls {@code createFn} exactly
+         * once with each {@code globalProcessorIndex} from 0 to {@code
+         * totalParallelism - 1}, this can help all the instances agree on which
+         * part of the data to emit.
+         *
+         * @param preferredLocalParallelism requested number of workers on each cluster member
+         */
+        public TimestampedStream<T> distributed(int preferredLocalParallelism) {
+            mPreferredLocalParallelism = preferredLocalParallelism;
+            return this;
+        }
+
+        /**
+         * Sets the limit on the amount of disorder (skew) present in the
+         * timestamps of the events this source will emit. Any given event's
+         * timestamp must be at most this many milliseconds behind the highest
+         * timestamp emitted so far, otherwise it will be considered late and
+         * dropped.
```

Comment:
might be good to link to `GeneralStage.addTimestamps`
it should also say that the disorder is within a single worker/processor only.